File: ros2/workspace/src/simple_ros/simple_ros/action_client.py
from asyncio import Future
import rclpy
from rclpy.node import Node
from rclpy.qos import (
    QoSDurabilityPolicy,
    QoSHistoryPolicy,
    QoSProfile,
    QoSReliabilityPolicy,
)
from std_msgs.msg import String
from std_srvs.srv import SetBool 
import time
import sys 
from rclpy.action import ActionClient
from rclpy.action.client import ClientGoalHandle
from action_msgs.msg import GoalStatus
from user_interface.action import Fibonacci

class Action_client(Node):
    def __init__(self):
        super().__init__("action_client")
        self.create_timer(1, self.print_hello)
        self.action_client = ActionClient(self, Fibonacci, "fibonacci")
        self.bool = bool()
        self.count = 0

    # 첫 서비스 실행 콜백 .. 
    def send_goal(self, step):
        self.get_logger().info(f"request goal : {step}")

        while not self.action_client.wait_for_server():
            self.count += 1
            self.get_logger().info(f"server not availible : {self.count}")
        goal = Fibonacci.Goal()
        goal.req_step = int(step)
        self.future = self.action_client.send_goal_async(goal, feedback_callback=self.feedback_callback )  
        self.future.add_done_callback(self.result_callback)
        
    
    def feedback_callback(self, msg):
        seq = msg.feedback.feedback_seq
        self.get_logger().info(f"feedback_seq : {seq}")

    def result_callback(self, future:Future):
        goal_handle : ClientGoalHandle= future.result()

        if not goal_handle.accepted:
            self.get_logger().info(f"goal is not accepted")
            return; 

        self.result_future = goal_handle.get_result_async()
        self.result_future.add_done_callback(self.get_result_callback)


    def get_result_callback(self, future:Future):
        goal_handle = future.result()
        status = goal_handle.status
        self.get_logger().info(f"get_result_callback status : {status}")
        result = goal_handle.result

        if status == GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info(f"get_result_callback result seq : {result.result_seq}")
        elif status == GoalStatus.STATUS_CANCELED:
            self.get_logger().info(f"get_result_callback result canceled : {result.result_seq}")
        elif status == GoalStatus.STATUS_CANCELING:
            self.get_logger().info(f"get_result_callback result canceling : {result.result_seq}")
        elif status == GoalStatus.STATUS_ABORTED:
            self.get_logger().info(f"get_result_callback result aborted : {result.result_seq}")
        elif status == GoalStatus.STATUS_ACCEPTED:
            self.get_logger().info(f"get_result_callback result accepted .. ")
        elif status == GoalStatus.STATUS_EXECUTING:
            self.get_logger().info(f"get_result_callback result executing ....")


    def print_hello(self):
        #self.get_logger().info(f"current bool : {self.bool}")
        pass

def main(args=None):
    rclpy.init(args=args)
    node = Action_client()
    try:
        node.send_goal(sys.argv[1])
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.destroy_node()

if __name__ == "__main__":
    main()

File: ros2/workspace/src/simple_ros_cpp/src/action_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "user_interface/action/fibonacci.hpp"
#include <chrono>
#include <iostream>
#include <thread>

using namespace std;
using namespace std::placeholders;
using namespace std::chrono_literals;

typedef user_interface::action::Fibonacci Fibonacci;
typedef rclcpp_action::ClientGoalHandle<Fibonacci> GoalHandleFibonacci;


class ActionClient : public rclcpp::Node
{
public:
    ActionClient()
        : Node("action_client")
    {
        _action_client = rclcpp_action::create_client<Fibonacci>(
            this, "fibonacci");
    }

    void send_goal(char* step) {
        while( !_action_client->wait_for_action_server(1s) ) {
            retry_cnt++; 
            RCLCPP_INFO(get_logger(), "can not connect server retry : %d", retry_cnt);
        }

        Fibonacci::Goal goal_msg = Fibonacci::Goal(); 
        auto options = rclcpp_action::Client<Fibonacci>::SendGoalOptions(); 
        options.goal_response_callback = bind(&ActionClient::goal_response_callback, this, _1); 
        options.feedback_callback = bind(&ActionClient::feedback_callback, this, _1, _2); 
        options.result_callback = bind(&ActionClient::result_callback, this, _1); 

        goal_msg.req_step = atoi(step); 
        _action_client->async_send_goal(goal_msg, options); 
    }

private:
    int retry_cnt = 0; 
    rclcpp_action::Client<Fibonacci>::SharedPtr _action_client;

    void goal_response_callback(
        const GoalHandleFibonacci::SharedPtr &goal_handle)
    {
        RCLCPP_INFO(get_logger(), "goal_response_callback"); 
        if( goal_handle ) {
            RCLCPP_INFO(get_logger(), "goal_response_callback goal_handle exist!!"); 
        }
    }

    void feedback_callback(
        const GoalHandleFibonacci::SharedPtr &goal_handle, 
        const shared_ptr<const Fibonacci::Feedback> feedback)
    {
        RCLCPP_INFO(get_logger(), "feedback_callback"); 
        goal_handle->get_goal_id(); 
        for(int seq : feedback->feedback_seq) {
            RCLCPP_INFO(get_logger(), "feedback_seq : %d", seq); 
        }
    }
    void result_callback(const GoalHandleFibonacci::WrappedResult &result)
    {
        RCLCPP_INFO(get_logger(), "result_callback"); 

        if( result.code == rclcpp_action::ResultCode::SUCCEEDED ) {
            
            for (int seq : result.result->result_seq) {
                RCLCPP_INFO(get_logger(), "seq : %d", seq);  
            } 
        } else if( result.code == rclcpp_action::ResultCode::CANCELED ) {
            RCLCPP_INFO(get_logger(), "result_callback => CANCELED"); 
        } 
        
    }

};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ActionClient>();
    node->send_goal(argv[1]); 

    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}